knowledge-base/grid-select-all-onread.md

```diff
@@ -0,0 +1,147 @@
+
+1. Use [`HeaderTemplate` for the `CheckboxColumn`](slug:components/grid/columns/checkbox#header-template) and add a custom [CheckBox component](slug:checkbox-overview) so you can have full control over its behavior.
+1. Handle the [`OnChange` event](slug:checkbox-events#onchange) of the CheckBox to track when the user checks/unchecks it to manage the selected items. 
+    - When the CheckBox is checked, request all the data from your datasource and assign it to the `SelectedItems` collection, so you have all items selected. You may want to cache all the data for further usage, so you don't need to request it additionally upon clicking the custom SelectAll CheckBox.
+    - When the user deselects the CheckBox, clear the  `SelectedItems` collection.
+1. Manage the [`Indeterminate` state](slug:checkbox-indeterminate-state) of the CheckBox based on the selected items' count.
+1. Track [when the user changes the `Indeterminate` state](slug:checkbox-events#indeterminatechanged) of the CheckBox (clicks the CheckBox when it is in `Indeterminate` state). In this case, ensure that the CheckBox value will be always set to true if you want to completely mimic the default CheckBox selection behavior.
+
+>caption Select all items with CheckBox when using OnRead
+
+````RAZOR
+@using Telerik.DataSource.Extensions
+
+<TelerikGrid TItem="@Employee"
+             OnRead="@ReadItems"
+             SelectionMode="@GridSelectionMode.Multiple"
+             @bind-SelectedItems="@SelectedEmployees"
+             Pageable="true">
+    <GridColumns>
+        <GridCheckboxColumn CheckBoxOnlySelection="true">
+            <HeaderTemplate>
+                @{
+                    <TelerikCheckBox @bind-Value="@SelectAllCheckBoxValue" 
+                                     Indeterminate="@(SelectedEmployees.Count() >= 1 && SelectedEmployees.Count() < Total)"
+                                     IndeterminateChanged="@HandleIndeterminateChanged"
+                                     OnChange="@HandleSelectAll"/>                    
+                }
+            </HeaderTemplate>
+        </GridCheckboxColumn>
+        <GridColumn Field=@nameof(Employee.ID) />
+        <GridColumn Field=@nameof(Employee.Name)/>
+        <GridColumn Field=@nameof(Employee.HireDate) Title="Hire Date" />
+    </GridColumns>
+</TelerikGrid>
+
+@code {
+    private List<Employee> SourceData { get; set; }
+
+    private IEnumerable<Employee> SelectedEmployees { get; set; } = Enumerable.Empty<Employee>();
+
+    private bool SelectAllCheckBoxValue { get; set; }
+
+    private int Total { get; set; } = 0;
+
+    //Track when the user checks/unchecks the custom SelectAll CheckBox
+    private void HandleSelectAll()
+    {
+        if (SelectAllCheckBoxValue)
+        {
+            // Request all the data and assign it to the SelectedItems collection. Here we use the already available SourceData for brevity.
+            SelectedEmployees = SourceData;
```

Comment:
Requesting all the data may not be possible or desirable, otherwise the Grid won't be using `OnRead`. Consider something like this instead: https://blazorrepl.telerik.com/mTaROTvI23ir7Eqy34 (from 1680863)